File: app/src/main/java/com/nikola/jakshic/dagger/view/fragment/HeroFragment.java
package com.nikola.jakshic.dagger.view.fragment;

import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.nikola.jakshic.dagger.DaggerApp;
import com.nikola.jakshic.dagger.diffcallback.HeroDiffCallback;
import com.nikola.jakshic.dagger.R;
import com.nikola.jakshic.dagger.util.NetworkUtil;
import com.nikola.jakshic.dagger.view.HeroSortDialog;
import com.nikola.jakshic.dagger.view.adapter.HeroAdapter;
import com.nikola.jakshic.dagger.viewModel.HeroViewModel;

import javax.inject.Inject;

/**
 * A simple {@link Fragment} subclass.
 */
public class HeroFragment extends Fragment implements HeroSortDialog.OnSortListener {

    @Inject
    ViewModelProvider.Factory viewModelFactory;
    private RecyclerView recyclerView;
    private HeroViewModel viewModel;
    private HeroAdapter mAdapter;
    private long accountId;

    public HeroFragment() {
        // Required empty public constructor
    }

    @Override
    public void onAttach(Context context) {
        ((DaggerApp) getActivity().getApplication()).getAppComponent().inject(this);
        super.onAttach(context);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_hero, container, false);

        SwipeRefreshLayout mRefresh = root.findViewById(R.id.swiperefresh_hero);

        accountId = getActivity().getIntent().getLongExtra("player-account-id", -1);

        viewModel = ViewModelProviders.of(this, viewModelFactory).get(HeroViewModel.class);

        recyclerView = root.findViewById(R.id.recview_hero);

        mAdapter = new HeroAdapter(getActivity(), new HeroDiffCallback());

        View btnSort = root.findViewById(R.id.btn_hero_sort);

        btnSort.setOnClickListener(v -> {
            HeroSortDialog dialog = HeroSortDialog.newInstance();
            dialog.setTargetFragment(this, 300);
            dialog.show(getFragmentManager(), null);
        });

        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        recyclerView.setAdapter(mAdapter);
        recyclerView.setHasFixedSize(true);

        viewModel.initialFetch(accountId);
        viewModel.getHeroes().observe(this, mAdapter::submitList);
        viewModel.getStatus().observe(this, status -> {
            switch (status) {
                case LOADING:
                    mRefresh.setRefreshing(true);
                    break;
                default:
                    mRefresh.setRefreshing(false);
                    break;
            }
        });


        mRefresh.setOnRefreshListener(() -> {
            if (NetworkUtil.isActive(getActivity())) {
                viewModel.fetchHeroes(accountId);
            } else {
                Toast.makeText(getActivity(), "Check network connection!", Toast.LENGTH_SHORT).show();
                mRefresh.setRefreshing(false);
            }
        });

        return root;
    }

    @Override
    public void onSort(int sortOption) {

        // Remove previous observers b/c we are attaching new LiveData<PagedList>
        viewModel.getHeroes().removeObservers(this);

        switch (sortOption) {
            case 0:
                viewModel.sortByGames(accountId);
                break;
            case 1:
                viewModel.sortByWinrate(accountId);
                break;
            case 2:
                viewModel.sortByWins(accountId);
                break;
            case 3:
                viewModel.sortByLosses(accountId);
                break;
        }

        // Set to null first, to delete all the items otherwise the list wont be scrolled to the first item
        mAdapter.submitList(null);

        // Attach the observer to the new LiveData<PagedList>
        viewModel.getHeroes().observe(this, mAdapter::submitList);
    }
}

File: app/src/main/java/com/nikola/jakshic/dagger/ui/profile/heroes/HeroFragment.kt
package com.nikola.jakshic.dagger.ui.profile.heroes

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.nikola.jakshic.dagger.*
import com.nikola.jakshic.dagger.ui.DaggerViewModelFactory
import com.nikola.jakshic.dagger.ui.Status
import kotlinx.android.synthetic.main.fragment_hero.*
import javax.inject.Inject

class HeroFragment : Fragment(), HeroSortDialog.OnSortListener {

    @Inject lateinit var factory: DaggerViewModelFactory
    private var id: Long = -1
    private lateinit var viewModel: HeroViewModel
    private lateinit var adapter: HeroAdapter

    override fun onAttach(context: Context?) {
        (activity?.application as DaggerApp).appComponent.inject(this)
        super.onAttach(context)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return container?.inflate(R.layout.fragment_hero)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewModel = ViewModelProviders.of(this, factory)[HeroViewModel::class.java]

        id = activity?.intent?.getLongExtra("account_id", -1) ?: -1

        viewModel.initialFetch(id)

        adapter = HeroAdapter()

        recView.layoutManager = LinearLayoutManager(context)
        recView.addItemDecoration(DividerItemDecoration(context, DividerItemDecoration.VERTICAL))
        recView.adapter = adapter
        recView.setHasFixedSize(true)

        viewModel.list.observe(this, Observer(adapter::addData))
        viewModel.status.observe(this, Observer {
            when (it) {
                Status.LOADING -> swipeRefresh.isRefreshing = true
                else -> swipeRefresh.isRefreshing = false
            }
        })

        val sortDialog = HeroSortDialog()
        sortDialog.setTargetFragment(this, 301)

        btnSort.setOnClickListener {
            if (!sortDialog.isAdded) sortDialog.show(fragmentManager, null)
        }

        swipeRefresh.setOnRefreshListener {
            if (hasNetworkConnection())
                viewModel.fetchHeroes(id)
            else {
                toast("Check network connection!")
                swipeRefresh.isRefreshing = false
            }
        }
    }

    override fun onSort(sort: Int) {
        // Remove previous observers b/c we are attaching new LiveData
        viewModel.list.removeObservers(this)

        when (sort) {
            0 -> viewModel.sortByGames(id)
            1 -> viewModel.sortByWinRate(id)
            2 -> viewModel.sortByWins(id)
            3 -> viewModel.sortByLosses(id)
        }
        // Set to null first, to delete all the items otherwise the list wont be scrolled to the first item
        adapter.addData(null)
        // Attach the observer to the new LiveData
        viewModel.list.observe(this, Observer(adapter::addData))
    }
}